server/src/main/java/org/opensearch/index/engine/InternalEngine.java

```diff
@@ -2156,15 +2148,26 @@ protected SegmentInfos getLatestSegmentInfos() {
      */
     @Override
     public GatedCloseable<SegmentInfos> getSegmentInfosSnapshot() {
-        final SegmentInfos segmentInfos = getLatestSegmentInfos();
+        final SegmentInfos segmentInfos;
+        final OpenSearchDirectoryReader reader;
+        final GatedCloseable<IndexCommit> lastIndexCommit = acquireLastIndexCommit(false);
         try {
-            indexWriter.incRefDeleter(segmentInfos);
+            reader = internalReaderManager.acquire();
+            SegmentInfos readerInfos = ((StandardDirectoryReader) reader.getDelegate()).getSegmentInfos();
+            // if the latest index commit on disk has a higher gen than the reader load that instead.
+            // This ensures our commit file will exist throughout upload process.
+            if (lastIndexCommit.get().getGeneration() > readerInfos.getGeneration()) {
```

Comment:
In what scenarios this `if` condition will be true? Flush created new commit but reader is not yet updated?

I am trying to understand if we discard newly created segment files by discarding `readerInfos`. This can result in data loss. Is it possible to add an assert to validate this?

Reply:
we can assert the SI versions are the same or that the commit is ahead.

This happens when:
1. index docs & refresh (R1) flushing new segments to disk and updating the reader.
2. During refresh listeners for R1 a new commit comes and wipes the segments_n on disk, replacing with segments_n+1
3. Still in refresh listeners we acquire a snapshot.  The reader still references segments_n while it is not on disk. (nosuchfile thrown in remote store listener).
4. Refresh R1 would complete - then Refresh R2 (blockign refresh after flush) will execute updating the reader to the latest infos but only if there are segment updates - didRefresh will return false here if this is an empty commit because nrt current check only considers the SI version not gen.


An alternative we could try here is to acquire the latest commit ref before refresh occurs in the refresh listener, and release it after refresh completes.  However, I think we could still at any point have a segments_n ref'd on the reader that is not on disk given flush can occur at any time.

Reply:
on point 4, if there are no segment updates, on a flush will it trigger a refresh? I did not follow the point completely. But are we saying that if R2 occurs on account of a flush (with no segment changes), then the readers info will not get updated and will continue to remain stale, while the latest commit has gone to next generation?